### PR TITLE
#46 Convert to DLL compatible plugin

### LINKED ISSUE #46: Math plugin could support CKEditor 5 DLL build
Currently math plugin must be added in CKEditor 5 build at compilation time. With new DLL build API it could be possible add math plugin afterwards.

Some documentation about DDL builds:
- https://ckeditor.com/docs/ckeditor5/27.1.0/builds/guides/development/dll-builds.html
- https://github.com/ckeditor/ckeditor5/issues/8395

### PR DESCRIPTION
**Pull request by DanielKulbe**
* DLL compatible plugin build process
* Fix some deprecations
* Update docs

closes #46

### CODE REVIEW

**Comment by tony** [comment]
Wow! This is impressive!

cc'ing @isaul32 as well

**Comment by isaul32** [comment]
This is a awesome feature. Could you @tony do the merge and release? Thanks very much.

**Comment by tony** [comment]
@DanielKulbe This will need another rebase (smaller bugfix to master)

**Comment by tony** [comment]
@DanielKulbe Give this a rebase?

**Comment by tony** [comment]
@DanielKulbe This is very good. wow!

Some issues we need to look at, consider the following diff:

Assuming `yarn start`:

## index.html

http://localhost:8080/

```diff
diff --git a/sample/index.html b/sample/index.html
index 8ef5e67..1b26c5a 100644
--- a/sample/index.html
+++ b/sample/index.html
@@ -30,6 +30,19 @@
 <h1>CKEditor 5 with ckeditor5-math – Development Sample</h1>

 <div id="editor">
+ <h2>The <code>Math</code> plugin <h2>
+    <p>
+        <code>Math</code> inserts mathematical formulas into the editor. You can click the CKEditor 5 Math icon in the toolbar and see the results.
+    </p>
+    <p><script type="math/tex">e=mc^2</script></p>
+    <p><script type="math/tex; mode=display">e=mc^2</script></p>
+    <p>
+       This should show "\test" as ≠ via katexRenderOptions.macros:
+       <script type="math/tex">\test</script>
+    </p>
+    <!-- Quill Style Tag -->
+    <p><span class="ql-formula" data-value="e=mc^2"></span></p>
+
     <h2>Development environment</h2>
     <p>
         This is a demo of the <a href="https://ckeditor.com/docs/ckeditor5/latest/builds/guides/overview.html#classic-editor">classic editor build</a> that loads the <code>Math</code> and <code>AutoformatMath</code> plugin.
```

The formulas don't display in the editor itself:

image

Preview works!

image

## dll.html

No editor showing for DLL builds

http://localhost:8080/dll.html

image

```
15Refused to execute script from '<URL>' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
dll.html:70 Uncaught ReferenceError: CKEditor5 is not defined
    at dll.html:70:56
```

**Comment by DanielKulbe** [comment]
Sorry, just had have time today to rebase.

Also addressed the issues above, But for the last one, I'm not sure what's happening on your end, because it just renders fine on my end:

image

**Comment by tony** [comment]
@DanielKulbe Thank you for the rebase and checking!

Sorry about this: I still can't get http://localhost:8080/dll.html to work

I wiped `node_modules/` and did `yarn && yarn` + `yarn start`

```
Uncaught Error: No element provided to render
    at s (auto-render.min.js:1:2575)
    at HTMLScriptElement.onload (dll.html:23:6)
15Refused to execute script from '<URL>' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
dll.html:68 Uncaught ReferenceError: CKEditor5 is not def
```

This is on:

```
node --version
v19.6.0
```

Is anyone else watching this getting http://localhost:8080/dll.html to work?

The main build at http://localhost:8080/ works fine.

**Comment by Wigny** [comment]
I don't know if that is the expected command to be run, but for me, just running `yarn dll:build` and later `yarn dll:serve` serves the DLL sample at http://127.0.0.1:8080/sample/dll.html as wanted...

File: sample/dll.html
Comment (Wigny) [inline sample/dll.html:19]: About the `Uncaught Error: No element provided to render` error that happens with Katex, I think it can be solved just by adding the `defer` attr to the script as described in [the doc](https://katex.org/docs/autorender.html#usage).

```suggestion
    <script
        defer
        src="https://cdn.jsdelivr.net/npm/katex@0.13.5/dist/contrib/auto-render.min.js"
```

Comment (DanielKulbe) [reply sample/dll.html:19]: I had to remove this to prevent load errors first reported.

Comment (DanielKulbe) [reply sample/dll.html:19]: But, hey ... feel free to try and report anything that works - I'm not really familiar with Katex, as I am using MathJax in my project.

**Comment by Wigny** [comment]
@isaul32, what is missing for we have this PR merged? Looking forward to using it on my projects...

**Comment by ghalestrilo** [comment]
I have to second what @Wigny said. Our CKEditor integration at work will be **so much simpler** once this is merged 🥲 
@tony can I kindly ask you to check this MR again? Thank you for your time.

**Comment by tony** [comment]
@ghalestrilo @Wigny @DanielKulbe I will try again tonight

**Comment by tony** [comment]
@ghalestrilo @Wigny Can you make a video of what steps you're taking to install + build and show what http://localhost:8080/dll.html looks like? 

<details>

<summary>Screenshot w/ inspector</summary>

image

</details>

This still doesn't work for me 🤷.

**Comment by Wigny** [comment]
@tony, here are my steps.

I think our different outputs on it are about the commands being run to serve the DLL sample page (as I think you are just running `yarn serve`). I don't think `ckeditor5-package-tools start` does the work, so I'm testing with `yarn dll:serve` instead... @DanielKulbe, can you please confirm which commands should be run to test the DLL build? Moreover, I think would be nice having these steps documented in the README for help in future contributions. 


https://user-images.githubusercontent.com/23202981/223465025-b39faf92-cf06-42a2-9d36-9c9b9b4991d9.mp4

**Comment by DanielKulbe** [comment]
Yes, as you statet it some comments above:

> I don't know if that is the expected command to be run, but for me, just running `yarn dll:build` and later `yarn dll:serve` serves the DLL sample at http://127.0.0.1:8080/sample/dll.html as wanted...

The whole process is already documented with the CKEditor package generator, I believe:
https://ckeditor.com/docs/ckeditor5/latest/framework/plugins/package-generator/javascript-package.html

**Comment by tony** [comment]
> yarn dll:serve

@Wigny Thank you, the approach [here](https://github.com/isaul32/ckeditor5-math/pull/83#issuecomment-1458354851) _does_ work for me.

- What **didn't** work was `yarn start` with http://127.0.0.1:8080/dll.html

  http://localhost:8080/node_modules/ckeditor5/build/ckeditor5-dll.js returns `Cannot GET /node_modules/ckeditor5/build/ckeditor5-dll.js`
- What **did** work was `yarn dll:serve` with http://127.0.0.1:8080/sample/dll.html

**Review by tony — APPROVED** [review]
@DanielKulbe Approved

- Tested locally with `yarn dll:serve` and `yarn start`
- Tested on a separate build

Well done, this is awesome

File: package.json
Comment (tony) [inline package.json:2]: @DanielKulbe I did find a minor issue here.

The npm package is ckeditor5-math

When corrected, it raises:

```
$ yarn run dll:build
yarn run v1.22.19
$ ckeditor5-package-tools dll:build
ckeditor5-math/node_modules/@ckeditor/ckeditor5-package-tools/lib/utils/get-webpack-config-dll.js:23
        const dllName = packageJson.name.split( '/' )[ 1 ].replace( /^ckeditor5-/, '' );
```

Comment (tony) [reply package.json:2]: I filed an issue at https://github.com/ckeditor/ckeditor5-package-generator/issues/139

**Comment by tony** [comment]
@isaul32 Are you okay with giving me publish access to @isaul32/ckeditor5-math on npm until https://github.com/ckeditor/ckeditor5-package-generator/issues/139 is handled?

This was a mistake on my part - it flew right past me in the review, I didn't expect it'd error with our normal package name.

File: package.json
Comment (tony) [reply package.json:2]: 100% an oversight by me 😆 Who would have thought an ordinary NPM package name would raise an error

**Comment by isaul32** [comment]
I'm okay with that. Do I have to do something?

**Comment by tony** [comment]
@isaul32 Yes, if you `git pull` master at [406aa5c](https://github.com/isaul32/ckeditor5-math/commit/406aa5cb3a4e2cabf4607e162aa861b57cabaac5), then run `yarn`, and `yarn publish --access public` - that should be enough to publish the NPM package to `@isaul32/ckeditor5-math`.

If that works - can you add me as a maintainer to the NPM package?

**Comment by isaul32** [comment]
I can do it this evening. I think you are already a maintainer of the NPM package.

image

**Comment by tony** [comment]
- [@isaul32/ckeditor5-math](https://www.npmjs.com/package/@isaul32/ckeditor5-math) doesn't exist yet - that's where the package will be published
- [ckeditor5-math](https://www.npmjs.com/package/ckeditor5-math) does exist, but can't be published due to https://github.com/ckeditor/ckeditor5-package-generator/issues/139

Thank you, you will be able to see this evening

**Comment by isaul32** [comment]
Done. I also invited you @tony again. https://www.npmjs.com/package/@isaul32/ckeditor5-math

**Comment by tony** [comment]
@isaul32 Thank you!

**Comment by tony** [comment]
@isaul32 Do you want to push the package.json with v36.0.4 to the git repo?

**Pull request merged**
